stl/inc/atomic

```diff
@@ -462,28 +475,30 @@ struct _Atomic_storage {
         const auto _Storage_ptr  = _STD addressof(_Storage);
         const auto _Expected_ptr = _STD addressof(_Expected);
 
-        for (;;) {
+        auto _Are_equal = [this, _Order](const void* _Comparand) -> bool {
             const _Ty _Observed      = load(_Order);
             const auto _Observed_ptr = _STD addressof(_Observed);
-            if (_CSTD memcmp(_Observed_ptr, _Expected_ptr, sizeof(_Ty)) != 0) {
+            if (_CSTD memcmp(_Observed_ptr, _Comparand, sizeof(_Ty)) == 0) {
+                return true;
+            }
 #if _CMPXCHG_MASK_OUT_PADDING_BITS
-                if constexpr (_Might_have_non_value_bits<_Ty>) {
-                    _Storage_for<_Ty> _Local;
-                    const auto _Local_ptr = _Local._Ptr();
-                    _CSTD memcpy(_Local_ptr, _Observed_ptr, sizeof(_Ty));
-                    __builtin_zero_non_value_bits(_Local_ptr);
-                    __builtin_zero_non_value_bits(_Expected_ptr);
-
-                    if (_CSTD memcmp(_Local_ptr, _Expected_ptr, sizeof(_Ty)) == 0) {
-                        _CSTD memcpy(_Expected_ptr, _Observed_ptr, sizeof(_Ty));
-                        continue;
-                    }
+            if constexpr (_Might_have_non_value_bits<_Ty>) {
+                _Storage_for<_Ty> _Local;
+                const auto _Local_ptr = _Local._Ptr();
+                _CSTD memcpy(_Local_ptr, _Observed_ptr, sizeof(_Ty));
+                __builtin_zero_non_value_bits(_Local_ptr);
+                __builtin_zero_non_value_bits(reinterpret_cast<_Ty*>(const_cast<void*>(_Comparand)));
+
+                if (_CSTD memcmp(_Local_ptr, _Comparand, sizeof(_Ty)) == 0) {
+                    _CSTD memcpy(const_cast<void*>(_Comparand), _Observed_ptr, sizeof(_Ty));

```

Comment:
This (and line 490) are writing through `_Comparand`, so I am confused as to why it's `const void*` in the first place. It looks like constness should be removed here, and this change propagated through the various callsites.

Reply:
Removed